components/eamxx/src/physics/mam/impl/mam4_amicphys.cpp

```diff
@@ -1611,159 +2185,172 @@ void modal_aero_amicphys_intr(
+
   // Saturation specific humidity
   const Real qv_sat = epsqs * ev_sat / (pmid - (1 - epsqs) * ev_sat);
 
-  const Real relhumgcm = haero::max(0.0, haero::min(1.0, qv / qv_sat));
+  // total # of subareas to do calculations for
+  int nsubarea;
 
-  // Set up cloudy/clear subareas inside a grid cell
-  int nsubarea, ncldy_subarea, jclea, jcldy;
+  // total # of cloudy subareas
+  int ncldy_subarea;
+
+  // indices of the clear and cloudy subareas
+  int jclea, jcldy;
+
+  // whether a subarea is cloudy
   bool iscldy_subarea[maxsubarea()];
+
+  // area fraction of each active subarea[unitless]
   Real afracsub[maxsubarea()];
+
+  // cloudy and clear fractions of the grid cell
+  Real fcldy, fclea;
+
+  setup_subareas(cld,                                      // in
+                 nsubarea, ncldy_subarea, jclea, jcldy,    // out
+                 iscldy_subarea, afracsub, fclea, fcldy);  // out
+  EKAT_KERNEL_ASSERT_MSG(nsubarea < maxsubarea(),
+                         "Error! modal_aero_amicphys_intr: "
+                         "nsubarea should be < maxsubarea() \n");
+
+  const Real relhumgcm = haero::max(0.0, haero::min(1.0, qv / qv_sat));
+
   Real relhumsub[maxsubarea()];
-  Real qsub1[gas_pcnst()][maxsubarea()];
-  Real qsub2[gas_pcnst()][maxsubarea()];
-  Real qsub3[gas_pcnst()][maxsubarea()];
-  Real qqcwsub1[gas_pcnst()][maxsubarea()];
-  Real qqcwsub2[gas_pcnst()][maxsubarea()];
-  Real qqcwsub3[gas_pcnst()][maxsubarea()];
-  // aerosol water mixing ratios (mol/mol)
-  Real qaerwatsub3[AeroConfig::num_modes()][maxsubarea()];
-  construct_subareas_1gridcell(cld, relhumgcm,                            // in
-                               q_pregaschem, q_precldchem,                // in
-                               qqcw_precldchem,                           // in
-                               q, qqcw,                                   // in
-                               nsubarea, ncldy_subarea, jclea, jcldy,     // out
-                               iscldy_subarea, afracsub, relhumsub,       // out
-                               qsub1, qsub2, qsub3,                       // out
-                               qqcwsub1, qqcwsub2, qqcwsub3, qaerwatsub3, // out
-                               qaerwat                                    // in
-  );
+  set_subarea_rh(ncldy_subarea, jclea, jcldy, afracsub, relhumgcm,  // in
+                 relhumsub);                                        // out
+
+  //-------------------------------
+  // Set aerosol water in subareas
+  //-------------------------------
+  // Notes from Dick Easter/Steve Ghan: how to treat aerosol water in
+  // subareas needs more work/thinking Currently modal_aero_water_uptake
+  // calculates qaerwat using the grid-cell mean interstital-aerosol
+  // mix-rates and the clear-area RH. aerosol water mixing ratios (mol/mol)
+  Real qaerwatsub3[num_modes][maxsubarea()];
+  assign_2d_array(num_modes, maxsubarea(), 0.0,  // in
```

Comment:
You can add `= {0}` above, and do away with this call. It's likely to be faster too.